File: 13.Roman-to-Integer.java
// https://leetcode.com/problems/roman-to-integer/
//
// algorithms
// Easy (52.45%)
// Total Accepted: 421,751
// Total Submissions: 804,113
// beats 66.39% of java submissions

class Solution {
    public int romanToInt(String s) {
        int[] values = new int[]{1, 5, 10, 50, 100, 500, 1000};
        char[] symbols = new char[]{'I', 'V', 'X', 'L', 'C', 'D', 'M'};
        HashMap<Character, Integer> map = new HashMap<>();
        
        for (int i = 0; i < 7; i++) {
            map.put(symbols[i], i);
        }
        
        int len = s.length();
        int idx = 0, res = 0;
        
        while (idx < len) {
            if (idx < len - 1) {
                int nowSymbolIdx = map.get(s.charAt(idx));
                int nextSymbolIdx = map.get(s.charAt(idx + 1));
                if (nowSymbolIdx >= nextSymbolIdx) {
                    res += values[nowSymbolIdx];
                    idx++;
                } else {
                    res += (values[nextSymbolIdx] - values[nowSymbolIdx]);
                    idx += 2;
                }
            } else {
                res += values[map.get(s.charAt(idx))];
                idx++;
            }
        }
        
        return res;
    }
}

File: 13.Roman-to-Integer.py
# https://leetcode.com/problems/container-with-most-water/
#
# algorithms
# Medium (44.59%)
# Total Accepted:    490,116
# Total Submissions: 919,349


class Solution(object):
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        length = len(s)
        if length == 0:
            return 0

        ch_2_value = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000,
        }
        res, idx = ch_2_value[s[0]], 1

        while idx < length:
            if (s[idx - 1] == 'I' and (s[idx] == 'V' or s[idx] == 'X')) or \
                    (s[idx - 1] == 'X' and (s[idx] == 'L' or s[idx] == 'C')) or \
                    (s[idx - 1] == 'C' and (s[idx] == 'D' or s[idx] == 'M')):
                res -= 2 * ch_2_value[s[idx - 1]]

            res += ch_2_value[s[idx]]
            idx += 1

        return res
